### ISSUE TITLE
[UX] Edit menu links form: reinstate the "expanded" checkbox column.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#2782]
## Describe your issue or idea

This feature was removed in https://www.drupal.org/node/564886, but I see no really valid reasons for it other than some comments about how some people thought that "it's not used as often" or that these columns ("enabled" and "expanded") are "in the way" (of what??). Unless there is a really valid usability issue that I'm missing, the removal of this column has made making menus expanded en mass a very cumbersome task (need to individually edit each link).

I propose that we reinstate this feature in Backdrop. We can use column `priority-medium` or `priority-low` classes to hide this column on mobile devices.

### Steps to reproduce (if reporting a bug)

Create a menu with 20+ items that have numerous children menu items (you can use devel_generate). Say you want to make some or all of the top-level menu items expanded for various block instances of the respective menu block (the default when you create menu items is to have the "Show as expanded" checkbox unticked)...

### Actual behavior (if reporting a bug)

1. Edit menu
2. Edit menu item
3. Check the "Show as expanded" checkbox
4. Save the form
5. Repeat steps 2 to 4 as many times as you have menu items that need to be expanded by default

With this method, doing this for 10 menu items requires 31 steps and 20 page refreshes.

### Expected behavior (if reporting a bug)

1. Edit menu
2. Immediately have an overview of how many menu items are set to be expanded by default or not (UX+)
3. Tick or untick checkboxes at will
4. Save the form

With this method, doing this for 10 menu items requires 12 steps and a single page refresh => UX++.

### Relevant version/system information (if applicable)

[Menu Expanded](https://www.drupal.org/project/menu_expanded) is a D7 module that reinstates the removed feature.

**Comment by quicksketch** [comment]
I'm up for revisiting this. Agreed that we have the option to hide these columns on small screen devices if needed. Explaining the "Expanded" option is always difficult. That may be the greatest challenge in restoring it to this page.

**Comment by klonos** [comment]
I think that we could:

a) use the same description/help text that this option has in the individual menu item edit form and add it above the table as a legend.

or b) introduce help icons (ℹ️ ) throughout the admin UI and have the help text be a pop-up. Then use that in this case to add the help icon next to the "Expanded" table header text.

**Comment by olafgrabienski** [comment]
Not sure if we need the checkbox anymore after #2773 came in Backdrop 1.8.0. What do you think, @klonos ?

**Comment by opi** [comment]
IMHO, this checkbox is still useful for dropdown menu style, or any code processing menu items

**Comment by klonos** [comment]
> Not sure if we need the checkbox anymore after #2773 came in Backdrop 1.8.0. What do you think, @klonos ?

I believe that #2773 solved the same problem, but in a different part of the admin UI. I had another case today where content admins of our site were confused about this option in the menu edit page (not the menu block). Also, config-wise, I think that this issue is saving the respective option per-menu-item, while #2773 saves it for the specific menu block.

Another thing to consider here is that although we would grant permissions to content admins to edit menus, we would *never* allow them to configure blocks/layouts.

So yes, I still believe that this is a very valid request.

I would hate to hold this useful UX improvement much longer. @quicksketch if we merge #3155, the explanation/description can live in an info message 😉

**Comment by BWPanda** [comment]
I also support this reinstatement. From what I understand, the 'expanded' option only affects drop-down menus, but even then I've had times where menus continue to drop-down even with this disabled on the parent (or maybe that was in Drupal...) So, what exactly does this do?

**Comment by BWPanda** [comment]
I've made a PR that adds the 'Expanded' checkboxes back to the menu overview pages: https://github.com/backdrop/backdrop/pull/3657

![image](https://user-images.githubusercontent.com/2385329/125068200-bf14c580-e0a4-11eb-9d16-279685611157.png)

**Comment by BWPanda** [comment]
And here's a similar PR, but this one only adds the checkboxes to 'parent' menu items: https://github.com/backdrop/backdrop/pull/3658

![image](https://user-images.githubusercontent.com/2385329/125071701-2b91c380-e0a9-11eb-9a5f-724ed07b3007.png)

The only problem is that when dragging menu items around, the checkboxes don't update until after you save:

![image](https://user-images.githubusercontent.com/2385329/125071887-685dba80-e0a9-11eb-89c3-7a9a6f6d3c63.png)

**Comment by indigoxela** [comment]
@BWPanda my personal preference would be to always add the checkbox. Although an item may not have children yet, they might get them. And possibly the items get added automatically and the users who trigger that, don't have permission to edit the menu. I see this as some sort of preparation for future menu items.

**Comment by klonos** [comment]
I agree with @indigoxela, for the same reasons.

**Comment by klonos** [comment]
I've reviewed and tested https://github.com/backdrop/backdrop/pull/3657 ...RTBC 👍🏼

**Comment by BWPanda** [comment]
Now I'm looking at the screenshots from my PRs, and I'm wondering if the 'Expanded' column should come *after* the 'Enabled' column, since 'Enabled' seems much more important...

**Comment by klonos** [comment]
I think that we should keep things as close as to how they were in D7, unless there's a valid UX issue.

**Comment by quicksketch** [comment]
> From what I understand, the 'expanded' option only affects drop-down menus, but even then I've had times where menus continue to drop-down even with this disabled on the parent (or maybe that was in Drupal...) So, what exactly does this do?

Elaborating on this further... the "Expanded" option does _not_ affect drop-downs. It only affects when the menu is displayed as a Hierarchical Tree, and even then it can be overridden by using the "Show all links" option:

![image](https://user-images.githubusercontent.com/100206/133524128-62b73eef-a922-4086-9cda-cddc5adc85db.png)

I do have a little hesitation adding these checkboxes again, as the expanded option really isn't used all that frequently. A part of me would prefer to remove the ability entirely and move it to contrib, but I don't think that's a feasible change for 1.x.

**Comment by philsward** [comment]
I'm going to go out on a limb and agree it should be left to contrib, that or maybe it would be worth adding an "extras" pack to core that would allow some things to be enabled and others disabled. This could be an excellent candidate for that idea.

Overall I'm good either way, but as @BWPanda pointed out, I think the columns should be enabled followed by expanded.

What about a "menu settings" page where this feature could be enabled or hidden depending on the needs of the builder? Kind of how Views has a settings page for various QoL stuff.

**Comment by klonos** [comment]
@quicksketch unless I'm horribly mistaken, nice_menus and superfish in D7 required that "Expanded" checkbox to be ticked for menu items that needed to be expanded when hovered-over (kinda makes sense, since you need the nested `<li>`s to be rendered). I'd need to test this on a D7 site to confirm though. ...perhaps it is not an issue because D7 creates its menu items with that option checked by default? ...I don't know 🤷🏼 

Regardless, if you need to enable the expanded option for multiple menu items, it becomes a UX nightmare to have to dig through each individual item, tick the box, save, repeat. And it's not just enabling/disabling that option - it's also a matter of getting an overview of which menu items are set to be expanded and which aren't. With that setting exposed in the menu item listing, this becomes a breeze: you get an instant over view, and you only need a single form submit to get things done.

**Comment by klonos** [comment]
> unless I'm horribly mistaken, nice_menus and superfish in D7 required that "Expanded" checkbox to be ticked for menu items that needed to be expanded when hovered-over...

I was horribly mistaken 😅

**Comment by bugfolder** [comment]
> And it's not just enabling/disabling that option - it's also a matter of getting an overview of which menu items are set to be expanded and which aren't. 

For this reason (and others), I very much like having the "expanded" checkboxes. (And I agree with "enabled followed by expanded.")

**Comment by indigoxela** [comment]
We just have to take care, that this doesn't conflict with any of the available settings for menu blocks - which provide a lot.

_Hierarchical tree_ has, for instance, a setting "Show all menu links", which expands all links regardless of their individual setting.
The _Dropdown menu_ ignores the expanded state in favor of its own Maximum depth setting, the _Top level only_ menu style does its own thing, anyway.

I think, we should clarify, if / why we really still need this setting. Don't panic, we probably can't remove it before 2.x.

**Comment by BWPanda** [comment]
After testing the 'expanded' functionality in different situations in https://github.com/backdrop/backdrop-issues/issues/5397, I've come to the same conclusion as @quicksketch in that I no longer think these checkboxes should be made more visible than they already are.

In Backdrop we can display menus in 3 different styles: top level only, dropdown menu, and hierarchial tree. The 'expanded' option for menu items only affects the hierarchial tree style, and even then only when the 'Show all menu links' option is disabled. So, all things being equal, that's like 17% of the time.

I therefore change my position on this matter to "close as 'won't fix'".

**Comment by indigoxela** [comment]
> I therefore change my position on this matter to "close as 'won't fix'".

@BWPanda I'm also fine with that.

@klonos as you opened this issue, maybe you want to consider the feedback given by several people here?

**Comment by klonos** [comment]
Yes, given the thorough explanation by @BWPanda, I'm convinced. I'm fine with closing this as "won't fix" or "works as designed" 👍🏼 

...wondering if there should be any documentation follow-ups though 🤔

**Comment by BWPanda** [comment]
Closing then. @klonos see https://github.com/backdrop/backdrop-issues/issues/5397

**Comment by quicksketch** [comment]
I don't think this has been suggested before, but it's possible we might be able to move the entire concept of "Expanded" from the menu item level to the block level. See the suggested approach I posted to https://github.com/backdrop/backdrop-issues/issues/5428.

**Comment by BWPanda** [comment]
For those still really wanting this, there is a port available: https://github.com/backdrop-contrib/menu_expanded
